File: tests/unit/generators/round.py
"""
Custom generators for rounding modes.
"""

import fpy2 as fp
from hypothesis import strategies as st

@st.composite
def rounding_modes(draw):
    """
    Returns a strategy for generating a rounding mode.
    """
    return draw(st.sampled_from([fp.RM.RNE, fp.RM.RNA, fp.RM.RTP, fp.RM.RTN, fp.RM.RTZ, fp.RM.RAZ, fp.RM.RTO, fp.RM.RTE]))

File: fpy2/cpp/src/round.cpp
#include "fpy/round.hpp"
#include "fpy/utils.hpp"

namespace fpy {

std::tuple<bool, RoundingDirection> to_direction(RoundingMode mode, bool sign) {
    switch (mode) {
        case RoundingMode::RNE:
            return { true, RoundingDirection::TO_EVEN };
        case RoundingMode::RNA:
            return { true, RoundingDirection::AWAY_ZERO };
        case RoundingMode::RTP:
            return { false, sign ? RoundingDirection::TO_ZERO : RoundingDirection::AWAY_ZERO };
        case RoundingMode::RTN:
            return { false, sign ? RoundingDirection::AWAY_ZERO : RoundingDirection::TO_ZERO };
        case RoundingMode::RTZ:
            return { false, RoundingDirection::TO_ZERO };
        case RoundingMode::RAZ:
            return { false, RoundingDirection::AWAY_ZERO };
        case RoundingMode::RTO:
            return { false, RoundingDirection::TO_ODD };
        case RoundingMode::RTE:
            return { false, RoundingDirection::TO_EVEN };
        default:
            FPY_UNREACHABLE("invalid rounding mode");
    }
}

} // namespace fpy
